tests/clickhouse-test

```diff
@@ -1441,18 +1629,51 @@ class TestCase:
             self.testcase_args = self.configure_testcase_args(
                 args, self.case_file, suite.suite_tmp_path
             )
+
             client_options = self.add_random_settings(client_options)
-            proc, stdout, stderr, debug_log, total_time = self.run_single_test(
-                server_logs_level, client_options
-            )
 
-            result = self.process_result_impl(
-                proc, stdout, stderr, debug_log, total_time
-            )
-            result.check_if_need_retry(args, stdout, stderr, self.runs_count)
-            # to avoid breaking CSV parser
-            result.description = result.description.replace("\0", "")
+            if not is_valid_utf_8(self.case_file) or not is_valid_utf_8(
+                self.reference_file
+            ):
+                proc, stdout, stderr, debug_log, total_time = self.run_single_test(
+                    server_logs_level, client_options
+                )
 
+                result = self.process_result_impl(
+                    proc, stdout, stderr, debug_log, total_time
+                )
+                result.check_if_need_retry(args, stdout, stderr, self.runs_count)
+                # to avoid breaking CSV parser
+                result.description = result.description.replace("\0", "")
+            else:
+                with SharedEngineReplacer(
+                    self.case_file,
+                    args.replace_replicated_with_shared,
+                    args.replace_non_replicated_with_shared,
+                    False,
+                ):
+                    with SharedEngineReplacer(
```

Comment:
Looks strange.

Reply:
This is just ugly (and to me usually hard to read) python code.  It just triggers thing in its destructor. I'm going to focus of having things passing in both repos with the same code first and then we can refactor things if necessary.